### ISSUE TITLE
Feature Request: allow other extensions to pass a URL to Save In

### ISSUE THREAD

**Issue by KernelLeak** [gyng/save-in#110]
I was trying to hook up Save In in Foxy Gestures by using `browser.sendMessage` with the optional `extensionId` parameter, only to find out that this works only when a listener for `browser.runtime.onMessageExternal` has been installed and not with the regular `browser.runtime.onMessage` handler... :(

Could you please add a handler for `onMessageExternal` so you can for example use Foxy Gestures' user script feature to pass the URL of an image to Save In via a `DOWNLOAD` message?

**Comment by gyng** [comment]
This is something I'm interested in doing. I'll try to get it done when I have some spare time.

**Comment by DanaMW** [comment]
I too am interested and am glad to help and to test attempts. I use uget with its official setup and follow your repo so here if you need me.

**Comment by gyng** [comment]
I've done a quick update to allow for this in 3.5.0, but I'm not a user of Foxy Gestures.

Take a look at https://github.com/gyng/save-in/wiki/Use-with-Foxy-Gestures and see if it works.

**Comment by tombones** [comment]
Thank you for adding this hook. I'm testing it out in Gesturefy, which appears to have similar "send message to other addon" integration:

![image](https://user-images.githubusercontent.com/3936058/58105568-b73bf380-7bb4-11e9-8aaf-a0cb3e5db2bd.png)

I'm no good at scripting, though. How would I need to modify your supplied sample code to just make this send a command to open the Save In menu? (assuming that's possible, which I suppose we'll find out together.)

**Comment by DanaMW** [comment]
Just noticed this upgrade and will test right now thanks for your work once again friend.

**Comment by KernelLeak** [comment]
Well, it does almost work. I'm sending the following payload via browser.runtime.sendMessage:

``` json
{
  "type": "DOWNLOAD",
  "body": {
    "url": "http://wapsisquare.com/wp-content/uploads/2019/05/nudge.jpg",
    "info": {
      "pageUrl": "http://wapsisquare.com/",
      "srcUrl": "http://wapsisquare.com/wp-content/uploads/2019/05/nudge.jpg",
      "comment": null
    }
  }
}
```

But then I get a toast stating "Save In: Failed to route or rename download" and it pops the standard Save As dialog, as expected when it can't route.

If I use the context menu's "Save In route..." it works as expected, so I'm pretty sure I've set up my routes correctly. Maybe the parameters I passed got mixed up along the line?

The code I used in Foxy Gestures was the following:

``` js
var source = data.element.mediaSource;

if (source) {
    const payload = {
        type: "DOWNLOAD",
        body: {
            url: source,
            info: { pageUrl: `${window.location}`, srcUrl: source, comment: null }
        }
    };

    browser.runtime.sendMessage("{72d92df5-2aa0-4b06-b807-aa21767545cd}", payload);
}
```

**Comment by KernelLeak** [comment]
Well, debugging the extension it seems that in `download.js` in line 57 in `getRoutingMatches` you expect the passed `info` object to have a `legacyDownloadInfo` member and pass that as the `info` object to `Router.matchRules`, but here it's actually undefined...

But since the parameter in `matchRules` is called `info` shouldn't the `info` object itself have been passed here, or at least if `legacyDownloadInfo` is null?

**Comment by gyng** [comment]
@KernelLeak thanks for taking the time to debug. I think it's got to do with a rule, since it works on a fresh install. If you are able to find out which rule is causing the problem and paste a version of it here it'll help me debug it.

![2019-05-22_02-44-30](https://user-images.githubusercontent.com/370496/58122209-ac787180-7c3b-11e9-8d52-2c6f0ca231a3.gif)

Anyway, I've put a quick untested release with `legacyDownloadInfo` ripped out at https://github.com/gyng/save-in/releases/tag/v3.5.1-a1, if you care to try untested code.

It's really confusing in that file because `info` is actually `legacyInfo`. I removed it once before last year but reverted it. (Or if you're adventurous enough I'm open to PRs...)

@tombones I tried poking around Gesturefy, but I don't see any way to run arbitrary JS or send dynamic messages to another addon. I'm only able to send static JSON strings, noting else. Maybe you can ask around on Gesturefy's support after this feature gets more ironed out.

**Comment by KernelLeak** [comment]
Well, since I'm not an add-on developer and haven't set my Firefox up for developing extensions I just took the 3.5.0 XPI, shoved the three changed JS files into it, disabled XPI signature checking in my Firefox Developer Edition and did a "Install extension from file..." - and now it works.

I'm pretty sure there must have been a deviation in both codepaths that made the original one work but not the new one, but I hope you don't mind me not trying to find out what it was exactly... :)

(I'm too busy watching our (Austrian) government fall apart right now to do more develop-y things...)

**Comment by gyng** [comment]
@KernelLeak thanks for the update. By "three files" did you mean the edited files from 3.5.1-a1?

**Comment by KernelLeak** [comment]
Yes, I meant the three files in the src folder that changed from 3.5.0 to 3.5.1-a1.

**Comment by tombones** [comment]
> [@tombones](https://github.com/tombones) I tried poking around Gesturefy, but I don't see any way to run arbitrary JS or send dynamic messages to another addon. I'm only able to send static JSON strings, nothing else. Maybe you can ask around on Gesturefy's support after this feature gets more ironed out.

Apologies for the thread necromancy @gyng and I hope you're well! If you're still working on this addon, it looks like Gesturefy now permits messages to be serialized as JS objects:

<img width="287" height="507" alt="Image" src="https://github.com/user-attachments/assets/b722d28f-4541-45df-86c8-e81f54488b73" />

I'd love to be able to set a gesture that directly opens the SaveIn menu. Any idea if that might be possible via message now?
